main.cpp

```diff
@@ -289,340 +241,164 @@ struct HeapA
 #include <functional>
 #include <memory>
 
-struct DoubleType;
-struct IntType;
 
-struct FloatType
+template<class T> 
+struct Numeric
 {
-    explicit FloatType(float lhs) : value( new float(lhs) ) {}
-    ~FloatType()
-    {
-        delete value;
-    }
+    using Type = T;
 
-    FloatType& operator+=(float rhs);
-    FloatType& operator-=(float rhs);
-    FloatType& operator*=(float rhs);
-    FloatType& operator/=(float rhs);
+    explicit Numeric(Type lhs) : value( std::make_unique<T>(lhs) ) {}
 
-    operator float() const
+    Numeric& operator+=(Type rhs)
     {
-        return *value;
+        *value += rhs; 
+        return *this;
     }
-
-    FloatType& pow(const FloatType&);
-    FloatType& pow(const DoubleType&);
-    FloatType& pow(const IntType&);
-    FloatType& pow(float);
-
-    FloatType& apply(std::function<FloatType&(float&)> myFunc);
-    FloatType& apply(void (* myFunc) (float&));
-
-
-private:
-    float* value = nullptr;
-    FloatType& powInternal(const float);
-
-};
-
-struct DoubleType
-{
-    explicit DoubleType(double lhs) : value( new double(lhs) ) {}
-    ~DoubleType()
+    Numeric& operator-=(Type rhs)
     {
-        delete value;
+        *value -= rhs; 
+        return *this;
     }
-
-    DoubleType& operator+=(double rhs);
-    DoubleType& operator-=(double rhs);
-    DoubleType& operator*=(double rhs);
-    DoubleType& operator/=(double rhs);
-
-    operator double() const 
+    Numeric& operator*=(Type rhs)
     {
-        return *value;
+        *value *= rhs; 
+        return *this;
     }
 
-    DoubleType& pow(const FloatType&);
-    DoubleType& pow(const DoubleType&);
-    DoubleType& pow(const IntType&);
-    DoubleType& pow(double);
-
-    DoubleType& apply(std::function<DoubleType&(double&)> myFunc);
-    DoubleType& apply(void (*myFunc) (double&));
-
-private:
-    double* value = nullptr;
-    DoubleType& powInternal(const double);
-
-};
-
-struct IntType
-{
-    explicit IntType(int lhs) : value(std::make_unique<int>(lhs)) {}
-
-    IntType& operator+=(int rhs);
-    IntType& operator-=(int rhs);
-    IntType& operator*=(int rhs);
-    IntType& operator/=(int rhs);
 
-    operator int() const
+    template<class ParamType>
+    Numeric& operator/=( ParamType rhs )
     {
-        return *value;
+        if constexpr (std::is_same<Type, int>::value)
+        {
+            if constexpr (std::is_same<ParamType, int>::value)
+            {
+                if(rhs == 0)
+                {
+                    std::cout << "error: integer division by zero is an error and will crash the program!\n";
+                    return *this;
+                }
+            }
+            else if (std::abs(rhs) <= std::numeric_limits<T>::epsilon())
+            {
+                std::cout << "can't divide integers by zero!\n";
+                return *this;
+            }
+        } 
+        else if (std::abs(rhs) <= std::numeric_limits<T>::epsilon())
```

Comment:
The instructions were:
```
                else if your parameter's value is less than epsilon
                        dont do the divison
        else if your parameter's value is less than epsilon
                warn about doing the division
```
You need to check if `rhs` is less than the epsilon of `rhs`'s type, not the Numeric's templated type.
- [ ] revise